Fix bug on close in NeoStores

davidegrohmann opened the pull request:
If any store does not close cleanly but throws an exception, the
subsequents stores in store array are not gonna be closed.  This will
make the page cache to fail when shutdown due to still store files
mapped in it.  This change will make sure we attempt to close all the
stores even if exceptions are thrown.

tinwelint commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/NeoStoresTest.java:
```diff
@@ -703,16 +705,54 @@ public void shouldNotSetHighestTransactionIdWhenNeeded() throws Throwable
+        try
+        {
+            long nextTxId = neoStore.getMetaDataStore().getLastCommittedTransactionId() + 1;
+            counts.rotate( nextTxId );
```
should you have a `fail(...)` here?

davidegrohmann replied:
we can add that

davidegrohmann replied:
done

tinwelint commented:
It's difficult to see the actual changes since 90% is checkstyle. For future commits, would it be possible to keep that separate?

chrisvest commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/NeoStoresTest.java:
```diff
@@ -107,8 +112,8 @@
-    public RuleChain ruleChain = RuleChain.outerRule( exception ).around( pageCacheRule )
-            .around( fs ).around( dir ).around( dsRule );
+    public RuleChain ruleChain =
+            RuleChain.outerRule( exception ).around( pageCacheRule ).around( fs ).around( dir ).around( dsRule );
```
I think the file system rule should be around the page cache rule, so we don't close the file system before we close the files.

davidegrohmann replied:
that's only formatting

chrisvest commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/NeoStoresTest.java:
```diff
@@ -485,14 +485,17 @@ public void setVersion() throws Exception
+                        Position.LOG_VERSION, 10 ) );
+        assertEquals( 10, MetaDataStore
+                .setRecord( pageCache, new File( storeDir, MetaDataStore.DEFAULT_NAME ).getAbsoluteFile(),
+                        Position.LOG_VERSION, 12 ) );
```
Not a fan of breaking at the dot so far in. Can you break at the start of the parameter list instead?

chrisvest commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/NeoStoresTest.java:
```diff
@@ -679,16 +681,16 @@ public void shouldSetHighestTransactionIdWhenNeeded() throws Throwable
-            store.setLastCommittedAndClosedTransactionId( 40, 4444, BASE_TX_COMMIT_TIMESTAMP,
-                    LogHeader.LOG_HEADER_SIZE, 0 );
+            store.setLastCommittedAndClosedTransactionId( 40, 4444, BASE_TX_COMMIT_TIMESTAMP, LogHeader.LOG_HEADER_SIZE,
+                    0 );
```
For cases such as this, I'd also prefer breaking at the start of the parameter list, or introduce a static import for the constant.

chrisvest commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/NeoStoresTest.java:
```diff
@@ -703,16 +705,54 @@ public void shouldNotSetHighestTransactionIdWhenNeeded() throws Throwable
+        try
+        {
+            // when we close the stores...
+            neoStore.close();
```
You also need a fail here, since you'd skip the asserts and checks in the catch clause if close doesn't throw.

davidegrohmann replied:
done

chrisvest commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/NeoStoresTest.java:
```diff
@@ -880,7 +916,7 @@ private int validateAndCountRelationships( long node, long rel1, long rel2, int
         int count = 0;
         try ( KernelStatement statement = (KernelStatement) tx.acquireStatement();
-              Cursor<NodeItem> nodeCursor = statement.getStoreStatement().acquireSingleNodeCursor( node ) )
+                Cursor<NodeItem> nodeCursor = statement.getStoreStatement().acquireSingleNodeCursor( node ) )
```
This extra indentation looks like a mistake. Resource declarations look better when they are aligned in a try-with-resources statement.

chrisvest commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/NeoStoresTest.java:
```diff
@@ -911,15 +947,15 @@ else if ( rel == rel2 )
     {
-        return ( property, propertyRecordId ) -> props.put( property.propertyKeyId(), Pair.of( property, propertyRecordId ) );
+        return ( property, propertyRecordId ) -> props
+                .put( property.propertyKeyId(), Pair.of( property, propertyRecordId ) );
```
Here I'd prefer to either break before `props`, or at the start of the parameter list.

chrisvest commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/NeoStoresTest.java:
```diff
@@ -962,7 +995,7 @@ else if ( data.propertyKeyId() == prop3.propertyKeyId() )
 
         try ( KernelStatement statement = (KernelStatement) tx.acquireStatement();
-              Cursor<NodeItem> nodeCursor = statement.getStoreStatement().acquireSingleNodeCursor( node ) )
+                Cursor<NodeItem> nodeCursor = statement.getStoreStatement().acquireSingleNodeCursor( node ) )
```
This one got indented as well.

chrisvest commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/NeoStoresTest.java:
```diff
@@ -1260,16 +1275,15 @@ private void assertHasRelationships( long node )
 
         try ( KernelStatement statement = (KernelStatement) tx.acquireStatement();
-              Cursor<NodeItem> nodeCursor = statement.getStoreStatement().acquireSingleNodeCursor( node ) )
+                Cursor<NodeItem> nodeCursor = statement.getStoreStatement().acquireSingleNodeCursor( node ) )
```
Another extra two-space indent.

chrisvest commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/NeoStoresTest.java:
```diff
@@ -1379,7 +1386,7 @@ private void testGetRels( long[] relIds )
     {
         try ( KernelStatement statement = (KernelStatement) tx.acquireStatement();
-              Cursor<NodeItem> nodeCursor = statement.getStoreStatement().acquireSingleNodeCursor( nodeId ) )
+                Cursor<NodeItem> nodeCursor = statement.getStoreStatement().acquireSingleNodeCursor( nodeId ) )
```
And another one.

chrisvest reviewed (changes requested):
The fix itself looks good, but the test can be improved and I think some of the stylistic changes can be improved as well.

davidegrohmann commented:
@chrisvest @tinwelint I'll revert the formatting mess

davidegrohmann committed:
Fix bug on close in NeoStores

If any store does not close cleanly but throws an exception, the
subsequents stores in store array are not gonna be closed.  This will
make the page cache to fail when shutdown due to still store files
mapped in it.  This change will make sure we attempt to close all the
stores even if exceptions are thrown.

davidegrohmann commented:
@chrisvest @tinwelint reverted the formatting changes, many comments were on code I didn't touch at all.

davidegrohmann commented:
@chrisvest @tinwelint any news on this PR?

chrisvest reviewed (approved):
Looks good.
